### ISSUE TITLE
Implement a migration script to migrate from 2.4 to 3.0

### ISSUE THREAD

**Issue by jrobinAV** [Avaiga/taipy-core#725]
Implement a script to migrate data for:
- All entities and versions
- All repository types

Ideally, the script should not depend on taipy library objects, so it is easy to maintain over time and to re-use for future taipy versions.

**Comment by joaoandre-avaiga** [comment]
Hey guys @jrobinAV @toan-quach @trgiangdo 

Can you take a look at the following script?
https://gist.github.com/joaoandre-avaiga/b513667985b53162c468402368840900

I have tested on a FileSystem repository, taking the entity changes from comparing old entities(from 2.0 and 2.4) and comparing with entities from 3.0. We could reuse most of this code to the other repositories as well. 

I do have a concern regarding SQL. I don't think we can manage to migrate data from SQL repository without depending on taipy core, because we need to know the data model of the database to be able to insert it. We can discuss other possible solutions as well.

If there is any thing you might remember missing, please let me know,

**Comment by trgiangdo** [comment]
Can we presume the data model for SQL? I think we can, right?

**Comment by jrobinAV** [comment]
The script does not include the Version repository.
Is it because there is nothing to migrate or because it is not handled yet?

**Comment by jrobinAV** [comment]
In line 124, when you check if the current entity must be migrated, you check if the entity prefix ("SCENARIO", "TASK", etc...) are part of the id. Right?
```
_entities = {k: data[k] for k in data if entity_type in k}
```
If I understand the code correctly, it may not work well. The `TASK` prefix is also included in a job id. And Scenario id can be part of a pipeline ID as well in some cases.

**Comment by jrobinAV** [comment]
> Hey guys @jrobinAV @toan-quach @trgiangdo
> 
> Can you take a look at the following script? https://gist.github.com/joaoandre-avaiga/b513667985b53162c468402368840900
> 
> I have tested on a FileSystem repository, taking the entity changes from comparing old entities(from 2.0 and 2.4) and comparing with entities from 3.0. We could reuse most of this code to the other repositories as well.
> 
> I do have a concern regarding SQL. I don't think we can manage to migrate data from SQL repository without depending on taipy core, because we need to know the data model of the database to be able to insert it. We can discuss other possible solutions as well.
> 
> If there is any thing you might remember missing, please let me know,

If we need to have the models in the migration code, should we have it for all repo type ? FS, SQL, Mongo ?
I never used it, but do you have experience with Alembic? It seems to be made for that purpose as well. No?

**Comment by toan-quach** [comment]
If we're talking about models, I think we can reuse the models we already have in core. Joao has written them in a way that they can be shared across repos within the need of rewriting it I believe 😃

**Comment by jrobinAV** [comment]
I have a concern in the structure of the migration script. Indeed to create a scenario 3.0, we need to have the scenario 2.4 itself with the attributes, plus some other entities like all the tasks of all the scenario s pipelines.

The structure the migrate scenario function cannot support that. Am I missing something?

```
def _migrate_scenario(scenario: Dict) -> Dict:
    # pipelines were replaced by tasks
    scenario["tasks"] = scenario["pipelines"]

    # pipeline attribute not removed in 3.0
    scenario["pipelines"] = None

    # additional_data_nodes attribute added
    scenario["additional_data_nodes"] = None

    return scenario
```

Besides this concern, I don't understand the scenario migration code itself, but that can be tuned or fixed later if needed.

**Comment by joaoandre-avaiga** [comment]
Let me address some of the points:
> The script does not include the Version repository.
Theres nothing to migrate from Version entity as far as I checked. The only thing is if they are coming from a taipy version where the Versioning system was not available(2.0 and bellow I think). This case is not handled by the script. If im not mistaken we used to handle this in taipy code itself. So for those on earlier versions may be needed to update to version 2.1 before going to 3.0. 

> If I understand the code correctly, it may not work well. The TASK prefix is also included in a job id. And Scenario id can be part of a pipeline ID as well in some cases.

Not the case, because all the entity names are capitalized. In the job example, `JOB_task_avg_div_5205c821-3d6e-45f3-b9a9-bcbeff92da6c` only the name of the entity is capitalized, everything else is in lower case, so there isn't an issue.

> If we need to have the models in the migration code, should we have it for all repo type ? FS, SQL, Mongo ?
I never used it, but do you have experience with Alembic? It seems to be made for that purpose as well. No?
Alembic is used to migrate the data model itself, as far as I have used before. We still can avoid using the methods by using raw SQL queries for the selects/inserts. My main point is that we won't be able to separate entirely from taipy, because the data model can be modified in future versions, so the script will need to be updated anyways. I prefer working with raw SQL in this case, makes it easier for the script. Also, I'll be using SQLAlchemy to perform the inserts, even with raw queries, but since it's already a dependency of taipy-core, I don't think it's an issue.

> I have a concern in the structure of the migration script. Indeed to create a scenario 3.0, we need to have the scenario 2.4 itself with the attributes, plus some other entities like all the tasks of all the scenario s pipelines.
Im not sure if I understand correctly what you meant. I'll try to explain my reasoning for the script structure. The scenario is the attribute that changed the least from 2.0 until 3.0. The only thing different on scenario is the change from `pipelines` to `tasks` which are handled by the first 2 lines in the function. The pipelines attribute is receiving a `None` because that is consistent with the scenario entities generated by taipy-core 3.0, maybe is an oversight and they need to be removed, but so far they are present. An example of a scenario 3.0 entity:
```
{
  "id": "SCENARIO_scenario_4c6c1a9e-d3d4-498a-8cde-b0b527e6c90c",
  "config_id": "scenario",
  "tasks": [
    "TASK_task_sum_excel_5bf595f4-47b2-4df1-8085-9a4b1f4a27af",
    "TASK_task_sum_csv_8dd2abbc-7d61-4bbe-a8f0-f9652dc343a2",
    "TASK_task_avg_mult_0da25a6e-541a-4c30-8976-84738e514e3f",
    "TASK_task_div_by_placeholder_18855981-f057-44ff-af66-bb85ee3e6cb7",
    "TASK_task_print_excel_f39d26f8-279f-4c5a-8f70-5ed8e80d4af4",
    "TASK_task_subtract_csv_excel_3f6cc2c0-eeb3-4032-9d19-b219ac4934c2",
    "TASK_task_mult_by_placeholder_1599859b-b275-408d-9de4-1d14b0802bcd",
    "TASK_task_insert_placeholder_56d58d83-069a-413c-9557-dd8ce676d57b",
    "TASK_task_print_csv_760c2c43-c87f-4eb2-8e6a-592afe9aad24",
    "TASK_task_avg_div_c40f5d58-6273-4361-add8-a4ab4fbee629"
  ],
  "additional_data_nodes": [],
  "properties": {},
  "creation_date": "2023-08-23T22:09:24.453989",
  "primary_scenario": false,
  "subscribers": [],
  "tags": [],
  "version": "3f8c81fc-756b-4625-9202-ff8962368761",
  "pipelines": null,
  "cycle": null
}
```
Also `additional_data_nodes` was added at some point, but Im not sure what values are it suppose to have, that's why is `None` but looking into it now maybe it should be an empty list instead, I think it makes more sense.


The script follows the following:
 - Migrate all the scenarios first. Because we need the tasks attribute to be filled in on scenario so we are able to fill in the tasks parent_ids, which is basically searching for all the scenarios that contains the task id for each task.
 - Migrate all the tasks. This is because: a) we have all the scenarios update and can fill in the parent_ids and b) because we need to fill in data node parent_ids as well. Which is the same as with tasks only looking into the tasks entities instead of scenarios.
 - Migrate all the datanodes. All the requirements are done, so we can go ahead and migrate the datanodes.
 - Migrate all the jobs. Here we also have the addition on of filed, `submit_entity_id`, Im not sure we are able to recover this information from the entities.
 

I've ignored pipelines so far, since they are being removed. Im not deleting any of them, but Im not changing any information on them. What do you think we should do with pipelines?

**Comment by jrobinAV** [comment]
Thanks a lot for the answer.

>  "Migrate all the scenarios first. Because we need the tasks attribute to be filled in on scenario so we are able to fill in the tasks parent_ids, which is basically searching for all the scenarios that contains the task id for each task."

OK so my understanding is that you filled the scenario.tasks attributes to migrate task.parent_id. Am I right? I thought it was to migrate pipeline.

The pipelines migration to scenario.sequences will also require replacing scenario.tasks by the concatenation of all scenario.pipelines[i].tasks. That why I was confused, but it is not implemented yet.

Sounds good to me then.
